component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java

```diff
@@ -0,0 +1,186 @@
+                if (e.getCause() instanceof CertificateException) {
+                    throw e;
+                } else {
+                    // This means certificate is not configured in service provider. In that case basic authentication
+                    // would be performed
+                    return super.authenticateClient(request, bodyParams, oAuthClientAuthnContext);
+                }
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticating client : " + oAuthClientAuthnContext.getClientId() + " with public " +
+                        "certificate.");
+            }
+
+            X509Certificate requestCert;
+            Object certObject = request.getAttribute(JAVAX_SERVLET_REQUEST_CERTIFICATE);
+            if (certObject instanceof X509Certificate[]) {
+                X509Certificate[] cert = (X509Certificate[]) certObject;
+                requestCert = cert[0];
+            } else if (certObject instanceof X509Certificate){
+                requestCert = (X509Certificate) certObject;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Could not find client certificate in required format for client: " +
+                            oAuthClientAuthnContext.getClientId());
+                }
+                return false;
+            }
+
+            return authenticate(registeredCert, requestCert);
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR, "Error occurred while retrieving " +
+                    "public certificate of client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (InvalidOAuthClientException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.INVALID_CLIENT, "Error occurred while retrieving " +
+                    "tenant domain for the client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR,
+                    "Unexpected error while authenticating client: " + oAuthClientAuthnContext.getClientId(), e);
+        }
+
+    }
+
+    public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
+                                   OAuthClientAuthnContext oAuthClientAuthnContext) {
+
+        // This authenticator will be skipped if BasicAuthClientAuthenticator was engaged
+        return !oAuthClientAuthnContext.getExecutedAuthenticators().contains(super.getName()) &&
+                super.canAuthenticate(request, bodyParams, oAuthClientAuthnContext);
+
+    }
+
+    public String getName() {
+        return "MutualTLSWithIdSecretAuthenticator";
+    }
+
+    /**
+     * Returns the execution order of this authenticator
+     *
+     * @return Execution place within the order
+     */
+    @Override
+    public int getPriority() {
+
+        return 101;
```

Comment:
Any idea why we need to hard code this priority ?

Reply:
We have given hardcoded getPriority() value 100 for BasicAuthClientAuthenticator (Extended class). If this value is not provided, same priority would be given to git authenticator.